### PR TITLE
Albumentations augs wrapper

### PR DESCRIPTION
**Pull request by tyomj**
Supporting aux augmentations from Albumentations (https://albumentations.readthedocs.io)

+  `Albu` transformation module as a wrapper
+  Example of usage based on MaskRCNN config file

### CODE REVIEW

**Comment by ternaus** [comment]
@hellock I am a developer of the Albumentation https://github.com/albu/albumentations image augmentation library and a big fan of the MMdetection.

Our library is well tested and supports numerous transformations on bounding boxes, masks, and images, allowing to change all 3 at the same time with a few lines of code.

In the last year, all or almost all winning solutions at Kaggle used it in their pipelines.

I believe that if we added its support to the mmdetection repo, it would significantly decrease the amount of code, related to preprocessing.

The winner of the Kaggle iMaterialist https://github.com/amirassov/kaggle-imaterialist challenged showed the value that albumentations may add to MMDetection.

Basically, if we would add the support of our library to mmdetection, it would make me happy :)

If some help is needed with the PR or we need to add some functionality to the Albumentation library, I would be happy to help.

**Comment by hellock** [comment]
Awesome! I am glad to integrate Albumentations into mmdetection, and will work on the review.
@tyomj Thanks for the efforts! There are some inconsistent code styles, e.g., we use single quote instead of double quotes, and do not add trailing commas.
@ternaus Thanks again for your nice package!

**Comment by tyomj** [comment]
@hellock Thanks for noticing! Fixed that.

**Comment by hellock** [comment]
@ternaus I briefly went through the library and it is really impressive! Of course this PR is a nice integration and I am glad to accept it. Furthermore, I wonder if there is a possibility to make closer collaboration. The APIs of Albumentation and MMDetection are somewhat similar and compatible, and Albumentation supports much more transforms. We may also consider switching to Albumentation as the default transform pipeline. Before that, we need to check the full compatibility and estimate the workload. Any ideas or comments?

**Comment by ternaus** [comment]
I agree with you. The APIs are similar. And I would be happy to help closer collaboration to happen. :)

If there are any changes in the albumentations that are required for deeper integration, my coauthors and I would be happy to do them.

**Comment by keyEpoch** [comment]
ValueError: Expected x_max for bbox [0.589461276405736, 0.2593853569030762, 1.2268104553222656, 0.5923876953125, 0] to be in the range [0.0, 1.0], got 1.2268104553222656.
Excuse me, must I regular x, y to [0.0, 1.0] before using albumetation in mmdetection?

**Comment by tyomj** [comment]
@keyEpoch No, it's not necessary. Albumentations performs bbox normalization based on the params you provide in `bbox_params` (see an example here `https://github.com/open-mmlab/mmdetection/pull/1354/files#diff-51670d2fae9c1b4f5fb4edd6dc32b2e2R167` and note that pipeline `LoadAnnotations` converts bbox coordinates to pascalvoc format even though you are using COCO-style dataset). 
So that the issue might be in params provided or in dataset directly. 
For more information about possible params and normalization please checkout the source: `https://github.com/albu/albumentations/blob/master/albumentations/augmentations/bbox_utils.py#L129`

**Comment by keyEpoch** [comment]
> @keyEpoch No, it's not necessary. Albumentations performs bbox normalization based on the params you provide in `bbox_params` (see an example here `https://github.com/open-mmlab/mmdetection/pull/1354/files#diff-51670d2fae9c1b4f5fb4edd6dc32b2e2R167` and note that pipeline `LoadAnnotations` converts bbox coordinates to pascalvoc format even though you are using COCO-style dataset).
> So that the issue might be in params provided or in dataset directly.
> For more information about possible params and normalization please checkout the source: `https://github.com/albu/albumentations/blob/master/albumentations/augmentations/bbox_utils.py#L129`


File "/home/perf/anaconda3/envs/mmdet_huan/lib/python3.7/site-packages/albumentations-0.3.2-py3.7.egg/albumentations/core/composition.py", line 156, in __call__
    p.preprocess(data)
  File "/home/perf/anaconda3/envs/mmdet_huan/lib/python3.7/site-packages/albumentations-0.3.2-py3.7.egg/albumentations/core/utils.py", line 66, in preprocess
    data[data_name] = self.check_and_convert(data[data_name], rows, cols, direction='to')
  File "/home/perf/anaconda3/envs/mmdet_huan/lib/python3.7/site-packages/albumentations-0.3.2-py3.7.egg/albumentations/core/utils.py", line 74, in check_and_convert
    return self.convert_to_albumentations(data, rows, cols)
  File "/home/perf/anaconda3/envs/mmdet_huan/lib/python3.7/site-packages/albumentations-0.3.2-py3.7.egg/albumentations/augmentations/bbox_utils.py", line 37, in convert_to_albumentations
    return convert_bboxes_to_albumentations(data, self.params.format, rows, cols, check_validity=True)
  File "/home/perf/anaconda3/envs/mmdet_huan/lib/python3.7/site-packages/albumentations-0.3.2-py3.7.egg/albumentations/augmentations/bbox_utils.py", line 188, in convert_bboxes_to_albumentations
    return [convert_bbox_to_albumentations(bbox, source_format, rows, cols, check_validity) for bbox in bboxes]
  File "/home/perf/anaconda3/envs/mmdet_huan/lib/python3.7/site-packages/albumentations-0.3.2-py3.7.egg/albumentations/augmentations/bbox_utils.py", line 188, in <listcomp>
    return [convert_bbox_to_albumentations(bbox, source_format, rows, cols, check_validity) for bbox in bboxes]
  File "/home/perf/anaconda3/envs/mmdet_huan/lib/python3.7/site-packages/albumentations-0.3.2-py3.7.egg/albumentations/augmentations/bbox_utils.py", line 148, in convert_bbox_to_albumentations
    check_bbox(bbox)
  File "/home/perf/anaconda3/envs/mmdet_huan/lib/python3.7/site-packages/albumentations-0.3.2-py3.7.egg/albumentations/augmentations/bbox_utils.py", line 214, in check_bbox
    value=value,
ValueError: Expected x_max for bbox [0.3653910273597354, 0.2511850396792094, 1.3556568509056455, 1.2275033791859944, 0] to be in the range [0.0, 1.0], got 1.3556568509056455.
File "/home/perf/anaconda3/envs/mmdet_huan/lib/python3.7/site-packages/albumentations-0.3.2-py3.7.egg/albumentations/augmentations/bbox_utils.py", line 214, in check_bbox
    value=value,
ValueError: Expected x_max for bbox [0.3653910273597354, 0.2511850396792094, 1.3556568509056455, 1.2275033791859944, 0] to be in the range [0.0, 1.0], got 1.3556568509056455.

Thanks for your reply. My dataset is coco format, so I changed the `bbox_params.format` to `coco`, other settings are the same as your default settings.  Errors occur when I use the default config settings. Mmdetection can run normally when I switch 'check_validity' to False `https://github.com/albu/albumentations/blob/master/albumentations/augmentations/bbox_utils.py#L175`. Does it matter or the bbox format used by albumentations must be normalized to [0, 1]? Could I fix it only in config file?

**Comment by tyomj** [comment]
> My dataset is coco format, so I changed the `bbox_params.format` to `coco`, other settings are the same as your default settings

Please, as I said earlier, use `pascal_voc` for `bbox_params.format` if u r loading annotation via mmdetectction coco dataset, because there is a transformation here:
https://github.com/open-mmlab/mmdetection/blob/master/mmdet/datasets/coco.py#L78-L81

That is why u probably have the error in the last two coordinates:
> ValueError: Expected x_max for bbox [0.3653910273597354, 0.2511850396792094, 1.3556568509056455, 1.2275033791859944, 0] to be in the range [0.0, 1.0], got 1.3556568509056455.

**Comment by ternaus** [comment]
I really need this PR to get merged for my projects, let me get involved :)

It is not clear what is happening.

It looks like it is something from the list:
1. Bug in albumentations.
2. Bug in this PR.
3. Issues with the annotations, say you have boxes that end outside of an image, or they are not in COCO format.

@keyEpoch

Could you please share a piece of the annotation that creates an error? I would like to reproduce it.

**Comment by keyEpoch** [comment]
Hi, my dataset is [lvis.](https://www.lvisdataset.org/), a coco format dataset. And I have addressed the problem( the bbox_params.format should be 'pascal_voc' for bbox format has been transformed already in 'LoadAnnotations' ).  But I met a new problem when I use 'ShiftScaleRotate', the error message is:
>
> > augmentations/bbox_utils.py", line 247, in check_bbox
    bbox=bbox,
ValueError: y_max is less than or equal to y_min for bbox [0.5390625, 0.6067708333333334, 0.541015625, 0.6067708333333334, 17].
>
And the error vanishes if I do not use 'ShiftScaleRotate'. The error is the same with #353
@ternaus

**Comment by ternaus** [comment]
@keyEpoch I looked at the labels => there are plenty of bounding boxes that go outside of the image. => issues with the annotation => feel free to fix provided noise in the dataset => everything will work.

@hellock It looks like it is the issue with annotations and not with the library.

What other steps do we need to merge this PR?

File: requirements.txt [hunk 1]
```diff
@@ -6,4 +6,5 @@ terminaltables
 pycocotools
 torch>=1.1
 torchvision
-imagecorruptions
\ No newline at end of file
+imagecorruptions
+albumentations==0.3.2
```
Comment (hellock) [inline requirements.txt]: We may just keep track of the latest albumentations.

**Comment by hellock** [comment]
@ternaus Yes I believe that the above issues are unrelated to the library. I will merge it and may open additional PRs for integrating the pipeline in the future.

File: requirements.txt (hunk 1 above)
Comment (tyomj) [reply requirements.txt]: Changed to >=

**Comment by DXist** [comment]
albumentations depends on `opencv-python-headless` opencv variant if opencv-python has not [already been installed](https://github.com/albu/albumentations/blob/master/setup.py#L35). But mmcv requires `opencv-python` variant so both are installed.

@hellock , how about changing  default mmcv dependency to `opencv-python-headless` as well?

**Comment by mosheliv** [comment]
+1

**Comment by 237014845** [comment]
@keyEpoch I have the same problem. Have you solved it？

ValueError: y_max is less than or equal to y_min for bbox (0.2289377643216041, 0.7070703125, 0.230590574202999, 0.7070703125, 5).

**Comment by 237014845** [comment]
@ternaus @tyomj I added albu_train_transforms from configs/albu_example/mask_rcnn_r50_fpn_1x.py to ATSS. Did this cause this error?

**Comment by openmmlab-v** [comment]
Hi @tyomj ！First of all, we want to express our gratitude for your significant PR in the mmdetection project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/raweFPmdzG

If you have WeChat，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤

**Comment by openmmlab-v** [comment]
Hi @tyomj ！First of all, we want to express our gratitude for your significant PR in the mmdetection project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/raweFPmdzG

If you have WeChat，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤
